p11-kit/list-objects.c

```diff
@@ -170,9 +235,17 @@ print_object (p11_list_printer *printer,
 	CK_MECHANISM_TYPE mechanism_type;
 	CK_BBOOL trusted, local, token, private, modifiable, copyable, destroyable;
 	CK_DATE start_date, end_date;
-	char label[128] = { '\0' };
-	char application[128] = { '\0' };
-	char id[128] = { '\0' };
+	char *uri_str;
+	size_t i;
+
+	CK_ATTRIBUTE arrs[] = {
```

Comment:
`arrs` is not very intuitive acronym I would say.

Reply:
Any suggestion? Otherwise I will go with `arr_attrs`.

Reply:
I would rather not split the attrs array, but just loop over the elements and if `pValue` is `NULL`, allocate memory as necessary, something like (without error handling):
```c
CK_ATTRIBUTE attrs[] = {
  { CKA_CERTIFICATE_CATEGORY, &cert_category, sizeof (cert_category) },
  ...
  { CKA_LABEL, NULL_PTR, 0 },
  ...
};

/* track allocated memory */
p11_array *allocated = p11_array_new (free);
...
/* first call to determine ulValueLen */
p11_kit_iter_get_attributes (iter, attrs, n_attrs);

for (i = 0; i < n_attrs; i++) {
  if (attrs[i].pValue == NULL && attrs[i].ulValueLen != CK_UNAVAILABLE_INFORMATION) {
    p11_buffer buf;
    size_t len = attrs[i].ulValueLen;

    p11_buffer_init (&buf, len);

    if (... this is a string attribute ...)
      p11_buffer_add (&buf, "\0", 1);

    attrs[i].pValue = p11_buffer_steal (&buf, &len);
    p11_array_push (allocated, attrs[i].pValue);
  }
}

/* second call to actually get the attributes */
p11_kit_iter_get_attributes (iter, arrs, n_arrs);

p11_array_free (allocated);
```